File: Languages/python/low-level-design-lld/design-patterns/singleton.py
"""
Singleton Design Pattern
Ensures a class has only one instance and provides a global point of access.

Thread-safe implementation using a metaclass.
"""

import threading


class SingletonMeta(type):
    """Thread-safe Singleton metaclass."""
    _instances = {}
    _lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            if cls not in cls._instances:
                instance = super().__call__(*args, **kwargs)
                cls._instances[cls] = instance
        return cls._instances[cls]


class Database(metaclass=SingletonMeta):
    def __init__(self):
        self.connection = "PostgreSQL@localhost:5432"

    def query(self, sql: str) -> str:
        return f"Executing: {sql}"


if __name__ == "__main__":
    db1 = Database()
    db2 = Database()

    assert db1 is db2, "Both should be the same instance"
    assert id(db1) == id(db2)
    assert db1.connection == "PostgreSQL@localhost:5432"
    assert db1.query("SELECT 1") == "Executing: SELECT 1"

    # Thread safety test
    results = []

    def create_instance():
        results.append(id(Database()))

    threads = [threading.Thread(target=create_instance) for _ in range(10)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    assert len(set(results)) == 1, "All threads should get the same instance"

    print("All tests passed!")

File: Languages/java/low-level-design-lld/design-patterns/Singleton.java
/**
 * Singleton Design Pattern
 * Thread-safe implementation using double-checked locking.
 */
public class Singleton {

    private static volatile Singleton instance;
    private final String connection;

    private Singleton() {
        this.connection = "PostgreSQL@localhost:5432";
    }

    public static Singleton getInstance() {
        if (instance == null) {
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }

    public String getConnection() {
        return connection;
    }

    public String query(String sql) {
        return "Executing: " + sql;
    }

    public static void main(String[] args) throws InterruptedException {
        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();

        assert s1 == s2 : "Should be same instance";
        assert s1.getConnection().equals("PostgreSQL@localhost:5432") : "Wrong connection";
        assert s1.query("SELECT 1").equals("Executing: SELECT 1") : "Query failed";

        // Thread safety test
        final Singleton[] instances = new Singleton[10];
        Thread[] threads = new Thread[10];
        for (int i = 0; i < 10; i++) {
            final int idx = i;
            threads[i] = new Thread(() -> instances[idx] = Singleton.getInstance());
            threads[i].start();
        }
        for (Thread t : threads) {
            t.join();
        }
        for (int i = 1; i < 10; i++) {
            assert instances[i] == instances[0] : "Thread safety violated";
        }

        System.out.println("All tests passed!");
    }
}
